yarn-project/archiver/src/archiver/kv_archiver_store/message_store.ts

```diff
@@ -1,220 +1,91 @@
-import { L1ToL2Message, NewInboxLeaf } from '@aztec/circuit-types';
-import { Fr, L1_TO_L2_MSG_SUBTREE_HEIGHT } from '@aztec/circuits.js';
+import { InboxLeaf } from '@aztec/circuit-types';
+import {
+  Fr,
+  INITIAL_L2_BLOCK_NUM,
+  L1_TO_L2_MSG_SUBTREE_HEIGHT,
+  NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP,
+} from '@aztec/circuits.js';
 import { createDebugLogger } from '@aztec/foundation/log';
-import { AztecCounter, AztecKVStore, AztecMap, AztecSingleton } from '@aztec/kv-store';
-
-/**
- * A message stored in the database
- */
-type Message = {
-  /** The L1ToL2Message */
-  message: Buffer;
-  /** The message's fee */
-  fee: number;
-  /** Has it _ever_ been confirmed? */
-  confirmed: boolean;
-};
+import { AztecKVStore, AztecMap, AztecSingleton } from '@aztec/kv-store';
 
 /**
  * LMDB implementation of the ArchiverDataStore interface.
  */
 export class MessageStore {
-  #newMessages: AztecMap<string, Buffer>;
-  #lastL1BlockNewMessages: AztecSingleton<bigint>;
-  // TODO(#4492): Nuke the following when purging the old inbox
-  #pendingMessagesByFee: AztecCounter<[number, string]>;
-  #messages: AztecMap<string, Message>;
-  #lastL1BlockAddingMessages: AztecSingleton<bigint>;
-  #lastL1BlockCancellingMessages: AztecSingleton<bigint>;
+  #l1ToL2Messages: AztecMap<string, Buffer>;
+  #l1ToL2MessageIndices: AztecMap<string, bigint>;
+  #lastL1BlockMessages: AztecSingleton<bigint>;
 
   #log = createDebugLogger('aztec:archiver:message_store');
 
   #l1ToL2MessagesSubtreeSize = 2 ** L1_TO_L2_MSG_SUBTREE_HEIGHT;
 
   constructor(private db: AztecKVStore) {
-    this.#newMessages = db.openMap('archiver_l1_to_l2_new_messages');
-    this.#messages = db.openMap('archiver_l1_to_l2_messages');
-    this.#pendingMessagesByFee = db.openCounter('archiver_messages_by_fee');
-    this.#lastL1BlockNewMessages = db.openSingleton('archiver_last_l1_block_new_messages');
-    this.#lastL1BlockAddingMessages = db.openSingleton('archiver_last_l1_block_adding_messages');
-    this.#lastL1BlockCancellingMessages = db.openSingleton('archiver_last_l1_block_cancelling_messages');
+    this.#l1ToL2Messages = db.openMap('archiver_l1_to_l2_messages');
+    this.#l1ToL2MessageIndices = db.openMap('archiver_l1_to_l2_message_indices');
+    this.#lastL1BlockMessages = db.openSingleton('archiver_last_l1_block_new_messages');
   }
 
   /**
    * Gets the last L1 block number that emitted new messages and the block that cancelled messages.
```

Comment:
Comment around cancelled messages.

Reply:
Addressed in 718447ef6bcb7f411e305e563e7534ea69fca0f7